Add Triton CPU as an Inductor backend

int3 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #133408

The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @rec

Differential Revision: [D63298968](https://our.internmc.facebook.com/intern/diff/D63298968)

int3 committed:
[prototype] Add Triton CPU as an Inductor backend

[ghstack-poisoned]

int3 committed:
Update on "[prototype] Add Triton CPU as an Inductor backend"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

int3 commented on torch/_inductor/config.py:
```diff
@@ -1032,7 +1032,7 @@ class rocm:
 
 # Backend to use for CPU codegen either "cpp" or "halide" (experimental)
-cpu_backend = "cpp"
+cpu_backend = "triton"
```
temporary while I test out stuff; will revert to the default before landing

int3 committed:
Update on "[prototype] Add Triton CPU as an Inductor backend"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

int3 commented on torch/_inductor/codegen/common.py:
```diff
@@ -1768,6 +1772,7 @@ def indirect_assert(
         if mask:
             cond = f"({cond}) | ~({mask})"
 
+        return ""  # skip emitting device_assert for now
```
also to be reverted before landing

int3 commented on torch/_inductor/graph.py:
```diff
@@ -1864,6 +1864,7 @@ def get_output_names(self) -> List[str]:
     def is_unspec_arg(self, name: str) -> bool:
         # dynamo wraps unspec variable as 0d CPU tensor,
         # need to convert to scalar during codegen (triton only)
+        return False  # FIXME
```
Looking for ideas on how to tackle this; I don't really understand why unspec variables need to be represented like so

int3 replied:
This was introduced in #87329; cc @jansel @yanboliang

desertfire replied:
This is because an unspecialized scalar is passed as a single-element cpu tensor input after dynamo. Should be ok to always return False for CPU-Triton.

int3 commented:
I plan to add some tests as well before landing, but would love to get some feedback on my overall approach first

desertfire commented:
> I plan to add some tests as well before getting a full review, but would love to get some feedback on my overall approach first

It will definitely make the review easier by breaking down into a stack of smaller PRs.

int3 commented:
> It will definitely make the review easier by breaking down into a stack of smaller PRs.

I think the changes in this PR look large because of indentation changes; it looks better when you ignore whitespace: https://github.com/pytorch/pytorch/pull/133408/files?w=1. But if you still think it's too large I can break it down further.

Anyway marking as draft for now while I fix things up

int3 committed:
Update on "[prototype] Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

int3 committed:
Update on "[prototype] Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

int3 committed:
Update on "[prototype] Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

int3 committed:
Update on "[prototype] Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

int3 committed:
Update on "[prototype] Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

int3 committed:
Update on "[prototype] Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

int3 committed:
Update on "[prototype] Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

int3 committed:
Update on "[prototype] Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

int3 committed:
Update on "[prototype] Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

int3 committed:
Update on "[prototype] Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

int3 committed:
Update on "[prototype] Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jansel commented:
Any early perf results for this?

int3 commented:
Not yet. Working on it...

int3 committed:
Update on "[prototype] Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

int3 committed:
Update on "[prototype] Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

int3 committed:
Update on "[prototype] Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

int3 committed:
Update on "[prototype] Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

int3 committed:
Update on "[prototype] Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

int3 committed:
Update on "[prototype] Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

int3 committed:
Update on "[prototype] Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

int3 committed:
Update on "[prototype] Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

int3 committed:
Update on "[prototype] Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

jansel commented on test/distributed/test_dynamo_distributed.py:
```diff
@@ -680,7 +693,9 @@ def test_fsdp_activation_checkpointing(self):
 
     @import_transformers_or_skip()
-    @unittest.skipIf(not has_triton(), "Inductor+gpu needs triton and recent GPU arch")
+    @unittest.skipIf(
+        not (has_triton() and HAS_GPU), "Inductor+gpu needs triton and recent GPU arch"
+    )
```
create a helper for this to avoid repeating it so many times

jansel commented on torch/_inductor/autotune_process.py:
```diff
@@ -646,19 +656,19 @@ def make_run_fn(
                 grid=self.grid,
                 **warmup_arg,
-                stream=get_raw_stream(self.output_tensor_meta.device.index),
+                stream=0,
```
?

jansel commented on torch/_inductor/autotune_process.py:
```diff
@@ -646,19 +656,19 @@ def make_run_fn(
         if "warmup" in inspect.signature(run_method).parameters:
             warmup_arg["warmup"] = False
 
-        from torch._C import _cuda_getCurrentRawStream as get_raw_stream
-
-        if torch.version.hip and self.matrix_instr_nonkdim != 0:
+        if output_tensor.device.type == "cpu":
```
?

int3 replied:
The two branches of the if/else were entirely identical, so I just repurposed one of them

int3 replied:
though I suppose it would be cleaner to factor out the `functools.partial` call and just use the branch for the value of `stream`

jansel commented on torch/_inductor/autotune_process.py:
```diff
@@ -646,19 +656,19 @@ def make_run_fn(
+                stream=0,
             )
         else:
+            from torch._C import _cuda_getCurrentRawStream as get_raw_stream
```
?

jansel commented on torch/_inductor/codegen/cpu/device_op_overrides.py:
This filename doesn't make much sense since most of the cpu stuff is one dir up

int3 replied:
feels like they should be moved in here to mirror the structure of the other backends. but I guess moving this file up is simpler

jansel replied:
I'm fine with a refactor, but that should be a different PR

int3 committed:
Update on "[prototype] Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

int3 committed:
Update on "[prototype] Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

int3 committed:
Update on "[prototype] Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

int3 commented on torch/_inductor/codegen/cpu_device_op_overrides.py:
```diff
@@ -0,0 +1,26 @@
+class CpuDeviceOpOverrides(DeviceOpOverrides):
+    def import_get_raw_stream_as(self, name):
+        return dedent(
+            """
+                      def get_raw_stream(_):
+                          return 0
+                      """
```
will clean up this formatting before landing

int3 commented on .ci/docker/build.sh:
```diff
@@ -379,6 +379,7 @@ case "$image" in
     GCC_VERSION=11
     CONDA_CMAKE=yes
     HALIDE=yes
+    TRITON=yes
```
I'm not really sure how `test_halide.py` was passing before, since it does seem to need triton: https://github.com/pytorch/pytorch/blob/f0ba309d78c8ab4930931afccb8afda63cce5ea1/test/inductor/test_halide.py#L224

jansel replied:
We are only running HalideCPU tests.

int3 commented:
bump -- this is ready for review

jansel commented:
Any results you can share?

int3 commented:
> Any results you can share?

I ran the following for 3 models and took the min timings from the raw metric dump:

```
TRITON_CPU_USE_SLEEF=~/local/sleef/build/lib/ TRITON_CPU_BACKEND=1 python benchmarks/dynamo/torchbench.py --inference --performance -dcpu -n30 --backend inductor --dump-raw-metrics --only ...
```

pytorch_unet triton / cpp time: 100.14%
pytorch_BERT triton / cpp time: 98.82%
timm_vision_transformer triton / cpp time: 105.72%

So it seems like it's roughly comparable

int3 commented on test/inductor/test_triton_cpu_backend.py:
```diff
@@ -0,0 +1,48 @@
+class TritonCpuTestMixin:
+    @classmethod
+    def setUpClass(cls):
+        os.environ["TRITON_CPU_BACKEND"] = "1"
```
It's used in the triton-cpu repo: https://github.com/triton-lang/triton-cpu/blob/ed82c339e5c9b9329d1ef3b3edea0d64e515f6ce/python/triton/runtime/driver.py#L8

int3 replied:
hmm, I guess it would be better if inductor set the env var globally if it knows the cpp backend is getting used. we can probably do that in the wrapper code. let me change this

int3 replied:
ah, nope, we would have to set it before doing any triton imports for e.g. autotuning as well, which gets kind of messy. I think in the longer run we should have a way to toggle the CPU backend without using env vars, but this is probably good enough for now

jansel replied:
Maybe inside inductor we should check for this and error if not set correctly?  I want users to have a good experience if they try to enable this.

int3 replied:
Sounds good, added an assert

desertfire commented:
> pytorch_unet triton / cpp time: 100.14% pytorch_BERT triton / cpp time: 98.82% timm_vision_transformer triton / cpp time: 105.72%
> 
> So it seems like it's roughly comparable

I have already added x86 to our OSS dashboard. You can submit a perf run at https://github.com/pytorch/pytorch/actions/workflows/inductor-perf-test-nightly-x86.yml, similar to the A100 one.

int3 commented:
hmm, I'll need to get the dashboard benchmarker to pull from https://github.com/triton-lang/triton-cpu instead of the regular triton repo though. I'll look into that later

int3 committed:
Update on "Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

WeizhuoZhang-intel commented:
Do people need to explicitly add a line `torch._inductor.config.cpp_backend=triton` in model script? Or we only need to `export TRITON_CPU_BACKEND=1`?

jansel commented on test/inductor/test_triton_cpu_backend.py:
Will this test file get run by CI?  I think we need to install the triton CPU backend on a runner and maybe add a test shard.

chunweiliu replied:
Hi Jason, is there any standard process to add a backend in Inductor?

int3 replied:
It won't, but a lot of the tests don't yet pass; the Triton CPU backend is still very much a WIP. I could filter out the subset that does pass, but could that be done in a future PR?

int3 replied:
bump

jansel replied:
Can we add the test file with a `@skipIfCpuTriton`?  I want to make sure it doesn't get broken over time.  This can be a followup PR in the stack though.

int3 replied:
okay this is done

int3 committed:
Update on "Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

int3 commented:
> Do people need to explicitly add a line torch._inductor.config.cpp_backend=triton in model script? Or we only need to export TRITON_CPU_BACKEND=1?

Both need to be done for now. I agree it's not a great situation though, we should have a better way of switching backends for Triton

int3 committed:
Update on "Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

int3 commented:
Results of the torchbench run, or what parts of it passed anyway, because the long compile times caused many jobs to time out...

<img width="1477" alt="Screenshot 2024-08-30 at 10 01 59 AM" src="https://github.com/user-attachments/assets/b261b561-25d5-4e71-a3e9-2cce9ea93b84">

So we are slightly behind in speed, but use about 2x the memory and 20x the compile time :/ lots of work to do here...

int3 committed:
Update on "Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

int3 committed:
Update on "Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

int3 committed:
Update on "Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

int3 committed:
Update on "Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

int3 committed:
Update on "Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

int3 committed:
Update on "Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

int3 commented on test/inductor/test_cuda_repro.py:
```diff
@@ -1315,7 +1315,9 @@ def fn(x, y):
         self.assertEqual(expect, actual)
 
         # Expect the code iterates in contiguous order, and is not tiled
-        kernel_code = "\n".join(code[0].split("\n")[60:74])
```
the line numbers changed since I added the new `set_driver_to_cpu()` call, so I've made the test a bit more flexible

int3 committed:
Update on "Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

int3 commented:
> Do people need to explicitly add a line torch._inductor.config.cpp_backend=triton in model script? Or we only need to export TRITON_CPU_BACKEND=1?

With the latest iteration of the PR, the user no longer needs to set `TRITON_CPU_BACKEND=1` :)

int3 committed:
Update on "Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

int3 committed:
Update on "Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

int3 committed:
Update on "Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

int3 committed:
Update on "Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

int3 commented:
@pytorchbot merge

jeanschmidt commented:
@pytorchbot revert -m "Broke internal signals, see D62737208 for more details" -c ghfirst

int3 committed:
Update on "Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

blaine-rister commented on torch/_inductor/runtime/triton_helpers.py:
```diff
@@ -31,6 +31,21 @@ def _log2(x):
+    if backend := triton.backends.backends.get("cpu", None):
+        triton.runtime.driver.set_active(backend.driver())
+        return
+    raise RuntimeError("Could not find an active CPU backend")
```
It seems like this is breaking out-of-tree backends like MTIA. It's possible for the user to create a new backend that PyTorch doesn't know about, and register it at runtime. If possible it would be better if we could be agnostic about which backend is in use, as opposed to erroring out if it's not one of N known backends.

Here are a few options that come to mind to work around the issue:
1. Remove the `raise RuntimeError` and make this a warning instead.
2. Refactor this into a method of `TritonScheduling`, so derived backends like MTIA can override it with their own `set_driver` function.
https://github.com/pytorch/pytorch/blob/main/torch/_inductor/codegen/triton.py#L2968

Does either of those sound reasonable?

blaine-rister replied:
By the way, we should probably add some tests for derived backends to catch this kind of thing in the future. I can look into that if folks agree.

int3 replied:
Per workchat conversation, I have gone with option #1. Since MTIA doesn't actually run the generated kernels on the CPU, it doesn't matter which active backend Triton has, so there's no need to error out.

blaine-rister replied:
Sounds good, thanks. Approved.

int3 committed:
Update on "Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

int3 commented:
@int3 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D63298968).

int3 committed:
Update on "Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

Differential Revision: [D63298968](https://our.internmc.facebook.com/intern/diff/D63298968)

[ghstack-poisoned]

int3 committed:
Update on "Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

Differential Revision: [D63298968](https://our.internmc.facebook.com/intern/diff/D63298968)

[ghstack-poisoned]

int3 committed:
Update on "Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

Differential Revision: [D63298968](https://our.internmc.facebook.com/intern/diff/D63298968)

[ghstack-poisoned]

int3 committed:
Update on "Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

Differential Revision: [D63298968](https://our.internmc.facebook.com/intern/diff/D63298968)

[ghstack-poisoned]

int3 committed:
Update on "Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

Differential Revision: [D63298968](https://our.internmc.facebook.com/intern/diff/D63298968)

[ghstack-poisoned]

int3 committed:
Update on "Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

Differential Revision: [D63298968](https://our.internmc.facebook.com/intern/diff/D63298968)

[ghstack-poisoned]

int3 commented:
@pytorchbot merge

int3 commented:
@pytorchbot merge

int3 commented:
@pytorchbot revert

int3 commented:
@pytorchbot --help

int3 commented:
@pytorchbot revert -m "internal tests failing" -c ghfirst

int3 commented:
@int3 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D63298968).

int3 committed:
Update on "Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

Differential Revision: [D63298968](https://our.internmc.facebook.com/intern/diff/D63298968)

[ghstack-poisoned]

int3 commented:
@int3 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D63298968).

int3 committed:
Update on "Add Triton CPU as an Inductor backend"


The goal is to use Inductor-generated kernels to stress test the new Triton CPU backend.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

Differential Revision: [D63298968](https://our.internmc.facebook.com/intern/diff/D63298968)

[ghstack-poisoned]

int3 commented:
@pytorchbot merge -f "test failure looks unrelated"
